model-optimizer/extensions/middle/MakeKaldiConstReshapable.py

```diff
@@ -0,0 +1,153 @@
+                init_value_prev_lstm_output = dest.node
+                break
+
+    if init_value_prev_lstm_output == {}:
+        fill_value = Const(graph, {'name': 'fill_value/'+input_name,
+                                   'value': np.array([value], dtype=precision),
+                                   'shape': int64_array([1])}).create_node()
+        init_value_prev_lstm_output = Broadcast(graph, {'name': 'init_value/'+input_name,
+                                                        }).create_node()
+        init_value_prev_lstm_output.in_port(0).connect(fill_value.out_port(0))
+        init_value_prev_lstm_output.in_port(1).connect(mem_shape.out_port(0))
+
+    return init_value_prev_lstm_output
+
+
+class MakeKaldiConstReshapable(MiddleReplacementPattern):
+    """
+    Add broadcasting of constant nodes based on batch from Parameter node. This approach works only for Kaldi,
+    because it has the same batch in whole graph due to framework specific.
+    """
+    enabled = True
+    graph_condition = [lambda graph: graph.graph['fw'] == "kaldi"]
+
+    def run_after(self):
+        from extensions.middle.InsertSelect import AddSelectBeforeMemoryNodePattern
+        from extensions.middle.ReplaceMemoryOffsetWithSplice import ReplaceMemoryOffsetWithMemoryNodePattern
+        from extensions.middle.ReplaceSpliceNodePattern import ReplaceSpliceNodePattern
+        return [AddSelectBeforeMemoryNodePattern, ReplaceMemoryOffsetWithMemoryNodePattern,
+                ReplaceSpliceNodePattern]
+
+    def find_and_replace_pattern(self, graph: Graph):
+        params = graph.get_op_nodes(op="Parameter")
+        batch = params[0].shape[0]
+
+        # check that all Parameters have the same batch
+        for p in params:
+            assert(p.shape[0] == batch)
+
+        # make constants for initialization of ReadValue reshapable
+        reads = graph.get_op_nodes(op='ReadValue')
+        for read in reads:
+            if read.in_port(0).get_source().node.op == "Const":
+                const = read.in_port(0).get_source().node
+                if len(const.out_port(0).data.get_shape()) != 2 or const.out_port(0).data.get_shape()[0] != batch:
+                    continue
+                new_const = create_const_with_batch_from_input(params[0].out_port(0),
+                                                               const.out_port(0).data.get_shape()[1],
+                                                               value=const.value[0], precision=const.data_type)
+                for dest in const.out_port(0).get_destinations():
+                    dest.get_connection().set_source(new_const.out_port(0))
+
+            # often we use ReadValue result to remove last with Crop and add first with Concat, new added values
+            # should have correct batch too
+            for dest in read.out_port(0).get_destinations():
```

Comment:
seems like a black magic) 
are there going to be other cases? should we split these into two transformations?

Reply:
I don't expect other cases, but in future of course anything can be changed. I don't want to split it in 2 transformations because we will need to duplicate checks and idea of both transformations is one.

Reply:
Agree with a comment about black magic. I would like to re-consider the whole approach of this change.

Reply:
Simplified transformation to add shapeof subgraphs for init values of ReadValue only